fabric-sponge-command-api-v2/src/main/java/net/fabricmc/fabric/api/command/v2/lib/sponge/args/GenericArguments.java

```diff
@@ -0,0 +1,2184 @@
+			}
+		}
+
+		@Override
+		protected Object parseValue(PermissibleCommandSource source, CommandArgs args) throws ArgumentParseException {
+			return args.hasNext() ? null : this.element.parseValue(source, args);
+		}
+
+		@Override
+		public List<String> complete(PermissibleCommandSource src, CommandArgs args, CommandContext context) {
+			return this.element.complete(src, args, context);
+		}
+
+		@Override
+		public Text getUsage(PermissibleCommandSource src) {
+			final Text containingUsage = this.element.getUsage(src);
+
+			if (containingUsage.getString().isEmpty()) {
+				return new LiteralText("");
+			}
+
+			return new LiteralText("[" + this.element.getUsage(src) + "]");
+		}
+	}
+
+	/**
+	 * Require a given command element to be provided a certain number of times.
+	 *
+	 * <p>Command values will be stored under their provided keys in the
+	 * <tt>CommandContext</tt>.</p>
+	 *
+	 * @param element The element to repeat
+	 * @param times   The number of times to repeat the element.
+	 * @return the element to match the input
+	 */
+	public static CommandElement repeated(CommandElement element, int times) {
+		return new RepeatedCommandElement(element, times);
+	}
+
+	private static class RepeatedCommandElement extends CommandElement {
+		private final CommandElement element;
+		private final int times;
+
+		protected RepeatedCommandElement(CommandElement element, int times) {
+			super(null);
+			this.element = element;
+			this.times = times;
+		}
+
+		@Override
+		public void parse(PermissibleCommandSource source, CommandArgs args, CommandContext context) throws ArgumentParseException {
+			for (int i = 0; i < this.times; ++i) {
+				this.element.parse(source, args, context);
+			}
+		}
+
+		@Override
+		protected Object parseValue(PermissibleCommandSource source, CommandArgs args) throws ArgumentParseException {
+			return null;
+		}
+
+		@Override
+		public List<String> complete(PermissibleCommandSource src, CommandArgs args, CommandContext context) {
+			for (int i = 0; i < this.times; ++i) {
+				CommandArgs.Snapshot startState = args.getSnapshot();
+
+				try {
+					this.element.parse(src, args, context);
+				} catch (ArgumentParseException e) {
+					args.applySnapshot(startState);
+					return this.element.complete(src, args, context);
+				}
+			}
+
+			return Collections.emptyList();
+		}
+
+		@Override
+		public Text getUsage(PermissibleCommandSource src) {
+			return new LiteralText(this.times + '*' + this.element.getUsage(src).getString());
+		}
+	}
+
+	/**
+	 * Require all remaining args to match as many instances of
+	 * {@link CommandElement} as will fit. Command element values will be stored
+	 * under their provided keys in the CommandContext.
+	 *
+	 * @param element The element to repeat
+	 * @return the element to match the input
+	 */
+	public static CommandElement allOf(CommandElement element) {
+		return new AllOfCommandElement(element);
+	}
+
+	private static class AllOfCommandElement extends CommandElement {
```

Comment:
This thing is an inner class nightmare, move all the impl classes to impl package

Reply:
blame api 7
inner classes good and best